### ISSUE TITLE
High Initial Latency Problem

### ISSUE THREAD

**Issue by liangfok** [liangfok/ros_shared_memory_interface#1]
I discovered a high initial latency problem and wrote a small example to demonstrate it. The demo consists of a:
- [publisher](https://github.com/liangfok/ros_shared_memory_interface/blob/feature/init_latency_test/shared_memory_interface_tutorials/src/tutorial_init_latency_test_publisher.cpp) - creates three publishers on three different topics, waits for the user to type [enter], then publishes a message on each topic at 1KHz
- [subscriber](https://github.com/liangfok/ros_shared_memory_interface/blob/feature/init_latency_test/shared_memory_interface_tutorials/src/tutorial_init_latency_test_subscriber.cpp) - subscribes to the publisher's messages, exists after receiving a message on each of the three topics.

To run the test:

```
$ roscore
$ rosrun shared_memory_interface shared_memory_manager
$ rosrun shared_memory_interface_tutorials tutorial_init_latency_test_subscriber
$ rosrun shared_memory_interface_tutorials tutorial_init_latency_test_publisher
```

The publisher prints the time when it starts to publish at 1KHz. The subscriber prints the times the first message is received on each topic. Here is example output:

```
[ INFO] [1434047912.386300170]: Publisher: Starting to publish at time 06-11-2015  13:38:32.386158

[ INFO] [1434047912.473625765]: Subscriber: callback1 called, time = 06-11-2015  13:38:32.473430, value = 87
[ INFO] [1434047912.474736670]: Subscriber: callback2 called, time = 06-11-2015  13:38:32.474640, value = 88
[ INFO] [1434047912.475969622]: Subscriber: callback3 called, time = 06-11-2015  13:38:32.475847, value = 89
```

The output indicates that the publisher started to publish at time 32.386158 and the subscriber received the first messages at times 32.473430, 32.474640, and 32.475847. This corresponds to initial latencies of 87.3ms, 88.5ms, and 89.7ms, respectively, which is very high. The values in the subscriber's output above are the number of times the publisher cycled before the subscriber received the first message. Note that the number of cycles matches the computed latencies since the publisher cycles at 1KHz.

**Comment by liangfok** [comment]
Update: I fixed a few bugs and optimized the test based on Josh's advice as follows:
1. Fixed a bug where the same subscriber was being used for all three topics.
2. Switched the subscribers to use non-polling mode, which is the default.
3. Modified publisher to only publish once.
4. Cache subscriber results and only print them after receiving all three messages.
5. Modified the publisher to publish once on all three topics before the benchmark begins. This is to more accurately model what's going on between ControlIt! and Gazebo.

The changes were saved in this commit: https://github.com/liangfok/ros_shared_memory_interface/commit/11b197de1cd59174fed5f6d7a61b915696266a2d

The results show significant reduction in the initial latency. The results below show latencies of around 0.2ms.

```
[ INFO] [1434056678.957328575]: Publisher: Starting to publish at time 06-11-2015  16:04:38.957230

[ INFO] [1434056678.957721375]: Subscriber: callback1 called, time = 06-11-2015  16:04:38.957424, value = 0
[ INFO] [1434056678.957780929]: Subscriber: callback2 called, time = 06-11-2015  16:04:38.957424, value = 0
[ INFO] [1434056678.957819832]: Subscriber: callback3 called, time = 06-11-2015  16:04:38.957451, value = 0
```

The results **are not significantly affected** when:
- The publisher continuously publishes at 1KHz.
- The publisher continuously publishes at 1KHz and the subscriber polls.

The results **are significantly affected** when the publisher does not publish on all topics prior to the start of the benchmark. An example output when the publisher publishes at 1KHz and the subscriber polls is shown below. It measures about 100ms of delay and indicates that the publisher transmits over 90 messages before the first one is received by the subscriber.

```
[ INFO] [1434058026.577799704]: Publisher: Starting to publish at time 06-11-2015  16:27:06.577632

[ INFO] [1434058026.674910891]: Subscriber: callback1 called, time = 06-11-2015  16:27:06.670701, value = 93
[ INFO] [1434058026.675362351]: Subscriber: callback2 called, time = 06-11-2015  16:27:06.671693, value = 94
[ INFO] [1434058026.675847908]: Subscriber: callback3 called, time = 06-11-2015  16:27:06.673725, value = 96
```

If the publisher only publishes once, the subscriber never receives the messages regardless of whether polling is used. The error case is saved in commit https://github.com/liangfok/ros_shared_memory_interface/commit/a50dc17e935373df059d6aeb9a8090e01d3777e7.

**Comment by liangfok** [comment]
I wrote a second pair of nodes to evaluate the initial latency when both nodes are publishers and subscribers. This more closely matches the relationship between a _controller_ and a _robot_. The demo consists of a:
- [robot](https://github.com/liangfok/ros_shared_memory_interface/blob/feature/init_latency_test/shared_memory_interface_tutorials/src/tutorial_init_latency_test_robot.cpp)
- [controller](https://github.com/liangfok/ros_shared_memory_interface/blob/feature/init_latency_test/shared_memory_interface_tutorials/src/tutorial_init_latency_test_controller.cpp)

The robot sends the controller data over three topics and the controller sends the robot data over two topics. The controller does not send anything until after it receives data from the robot across all three topics.

The robot and controller both contain a C pre-processor macro called "PRE_PUBLISH" that specifies whether messages are pre-published on each topic prior to the benchmark.

To run the test:

```
$ roscore
$ rosrun shared_memory_interface shared_memory_manager
$ rosrun shared_memory_interface_tutorials tutorial_init_latency_test_controller
$ rosrun shared_memory_interface_tutorials tutorial_init_latency_test_robot
```

Below are the results of a test in which messages are **not pre-published** and both the robot and controller **publish at 1KHz**. They show that there is about a 25ms delay in transmitting initial data from the robot to the controller, and a 61ms delay in transmitting initial data from the controller to the robot.

```
[ INFO] [1434113358.534656477]: Robot: Starting to publish at time 06-12-2015  07:49:18.534503
[ INFO] [1434113359.537604326]: Robot: callback1 called, time = 06-12-2015  07:49:18.622407, value = 61
[ INFO] [1434113359.537871601]: Robot: callback2 called, time = 06-12-2015  07:49:18.622681, value = 61

[ INFO] [1434113358.561633704]: Controller: Starting to publish at time 06-12-2015  07:49:18.561391
[ INFO] [1434113359.561568119]: Controller: callback1 called, time = 06-12-2015  07:49:18.558732, value = 24
[ INFO] [1434113359.561720555]: Controller: callback2 called, time = 06-12-2015  07:49:18.560158, value = 25
[ INFO] [1434113359.561876061]: Controller: callback3 called, time = 06-12-2015  07:49:18.561376, value = 26
```

Below are the results of a test in which messages are **pre-published** and both the robot and controller **publish at 1KHz**. They show that there is about a 0.3ms delay in transmitting initial data from the robot to the controller, and a 0.2ms delay in transmitting initial data from the controller to the robot.

```
[ INFO] [1434113241.911171547]: Robot: Starting to publish at time 06-12-2015  07:47:21.910998
[ INFO] [1434113242.911220142]: Robot: callback1 called, time = 06-12-2015  07:47:21.911817, value = 0
[ INFO] [1434113242.911323526]: Robot: callback2 called, time = 06-12-2015  07:47:21.911888, value = 0

[ INFO] [1434113241.911726178]: Controller: Starting to publish at time 06-12-2015  07:47:21.911623
[ INFO] [1434113242.911723403]: Controller: callback1 called, time = 06-12-2015  07:47:21.911287, value = 0
[ INFO] [1434113242.911823867]: Controller: callback2 called, time = 06-12-2015  07:47:21.911305, value = 0
[ INFO] [1434113242.911895229]: Controller: callback3 called, time = 06-12-2015  07:47:21.911314, value = 0
```

Below are the results of a test in which messages are **not pre-published** and both the robot and controller **publish once**. They show that there is about a 56ms delay in transmitting initial data from the robot to the controller, and a 83ms delay in transmitting initial data from the controller to the robot.

```
[ INFO] [1434114411.623967560]: Robot: Starting to publish at time 06-12-2015  08:06:51.623810
[ INFO] [1434114411.767098443]: Robot: callback1 called, time = 06-12-2015  08:06:51.766366, value = 0
[ INFO] [1434114411.767233459]: Robot: callback2 called, time = 06-12-2015  08:06:51.766585, value = 0

[ INFO] [1434114411.683128633]: Controller: Starting to publish at time 06-12-2015  08:06:51.682962
[ INFO] [1434114411.684071575]: Controller: callback1 called, time = 06-12-2015  08:06:51.679858, value = 0
[ INFO] [1434114411.684155872]: Controller: callback2 called, time = 06-12-2015  08:06:51.681171, value = 0
[ INFO] [1434114411.684200066]: Controller: callback3 called, time = 06-12-2015  08:06:51.682366, value = 0
```

Below are the results of a test in which messages are **pre-published** and both the robot and controller **publish once**. They show that there is about a 0.2ms delay in transmitting initial data from the robot to the controller, and a 0.2ms delay in transmitting initial data from the controller to the robot.

```
[ INFO] [1434114483.287471116]: Robot: Starting to publish at time 06-12-2015  08:08:03.287327
[ INFO] [1434114483.288499554]: Robot: callback1 called, time = 06-12-2015  08:08:03.288066, value = 0
[ INFO] [1434114483.288570405]: Robot: callback2 called, time = 06-12-2015  08:08:03.288093, value = 0

[ INFO] [1434114483.287968247]: Controller: Starting to publish at time 06-12-2015  08:08:03.287869
[ INFO] [1434114483.288950316]: Controller: callback1 called, time = 06-12-2015  08:08:03.287567, value = 0
[ INFO] [1434114483.289025273]: Controller: callback2 called, time = 06-12-2015  08:08:03.287572, value = 0
[ INFO] [1434114483.289074932]: Controller: callback3 called, time = 06-12-2015  08:08:03.287586, value = 0
```

The results show it is very important to pre-publish a message on each topic prior to the start of the benchmark. They also show you don't need to publish multiple times for the first message to get through regardless of whether messages are pre-published.

**Comment by liangfok** [comment]
I modified tutorial_init_latency_test_subscriber and tutorial_init_latency_test_publisher to contain logic that is similar to tutorial_init_latency_test_robot and tutorial_init_latency_test_controller. Here is the commit: https://github.com/liangfok/ros_shared_memory_interface/commit/3ec400b88f0914aed5eb11bbc1c56ef305682762

The results when messages are **not pre-published** and only one message is sent are shown below. They indicate initial latencies of about 60ms.

```
[ INFO] [1434125753.731559005]: Publisher: Starting to publish at time 06-12-2015  11:15:53.731457
[ INFO] [1434125753.797826859]: Subscriber: callback1 called, time = 06-12-2015  11:15:53.793647, value = 0
[ INFO] [1434125753.797929037]: Subscriber: callback2 called, time = 06-12-2015  11:15:53.795042, value = 0
[ INFO] [1434125753.798013265]: Subscriber: callback3 called, time = 06-12-2015  11:15:53.796757, value = 0
```

The results when messages are **pre-published** and only one message is sent are shown below. They indicate initial latencies of about 0.2ms.

```
[ INFO] [1434126134.988281185]: Publisher: Starting to publish at time 06-12-2015  11:22:14.988180
[ INFO] [1434126134.989476170]: Subscriber: callback1 called, time = 06-12-2015  11:22:14.988415, value = 0
[ INFO] [1434126134.989554183]: Subscriber: callback2 called, time = 06-12-2015  11:22:14.988416, value = 0
[ INFO] [1434126134.989613478]: Subscriber: callback3 called, time = 06-12-2015  11:22:14.988416, value = 0
```

In conclusion, pre-transmitting a message on each topic is very important to prevent high initial latency. For the benchmark, just one message needs to be transmitted regardless of whether a message is pre-transmitted on each topic.

**Comment by liangfok** [comment]
When using the shard memory interface with Gazebo, a node that initially fails to connect to a shared memory topic will not connect until the simulation begins. This is because the thread establishes the connection loops using `ros::Rate`, which by default does not advance when the simulation time remains stopped. To fix this problem, I switched to use `usleep(...)` instead of `ros::Rate` during the connection stage. Here is the commit: https://github.com/liangfok/ros_shared_memory_interface/commit/7614cb6e152cad48062597fc1839561f8f502423

By making this switch, the initial high latency caused by the need to connect after the simulation begins is solved.
